bundles/org.openhab.binding.teleinfo/src/main/java/org/openhab/binding/teleinfo/internal/serial/TeleinfoSerialControllerHandler.java

```diff
@@ -0,0 +1,195 @@
+        logger.warn("Retry in progress. Next retry in {} seconds...", SERIAL_PORT_DELAY_RETRY_IN_SECONDS);
+        updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.NONE, ERROR_UNKNOWN_RETRY_IN_PROGRESS);
+        try {
+            Thread.sleep(SERIAL_PORT_DELAY_RETRY_IN_SECONDS * 1000);
+            return true;
+        } catch (InterruptedException e1) {
+            return false;
+        }
+    }
+
+    private void openSerialPortAndStartReceiving() {
+        TeleinfoSerialControllerConfiguration config = getConfigAs(TeleinfoSerialControllerConfiguration.class);
+
+        if (config.serialport.trim().isEmpty()) {
+            logger.warn("Teleinfo port is not set.");
+            return;
+        }
+
+        logger.debug("Connecting to serial port '{}'...", config.serialport);
+        String currentOwner = null;
+        try {
+            final SerialPortIdentifier portIdentifier = serialPortManager.getIdentifier(config.serialport);
+            logger.debug("portIdentifier = {}", portIdentifier);
+            if (portIdentifier == null) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                        ERROR_OFFLINE_SERIAL_NOT_FOUND);
+                return;
+            }
+            logger.debug("Opening portIdentifier");
+            currentOwner = portIdentifier.getCurrentOwner();
+            logger.debug("portIdentifier.getCurrentOwner() = {}", currentOwner);
+            SerialPort commPort = portIdentifier.open("org.openhab.binding.teleinfo", 5000);
+            serialPort = commPort;
+
+            commPort.setSerialPortParams(1200, SerialPort.DATABITS_7, SerialPort.STOPBITS_1, SerialPort.PARITY_EVEN);
+            commPort.enableReceiveThreshold(1);
+            commPort.enableReceiveTimeout(SERIAL_RECEIVE_TIMEOUT);
+            logger.debug("Starting receive thread");
+            TeleinfoReceiveThread receiveThread = new TeleinfoReceiveThread(commPort, this,
+                    config.autoRepairInvalidADPSgroupLine, scheduler);
+            this.receiveThread = receiveThread;
+            receiveThread.start();
+
+            logger.debug("Connected to serial port '{}'", config.serialport);
+        } catch (PortInUseException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                    ERROR_OFFLINE_SERIAL_INUSE);
+        } catch (UnsupportedCommOperationException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                    ERROR_OFFLINE_SERIAL_UNSUPPORTED);
+        }
+    }
+
+    private void stopReceivingAndCloseSerialPort() {
+        TeleinfoReceiveThread receiveThreadRef = receiveThread;
+        if (receiveThreadRef != null) {
+            receiveThreadRef.interrupt();
+            try {
+                receiveThreadRef.join();
```

Comment:
You might want to add a timeout parameter to `join()` in case something blocks the interrupt.